engine/crates/integration-tests/tests/federation/basic/variables.rs

```diff
@@ -44,42 +48,398 @@ fn lists() {
+    ]
+    "###);
+    insta::assert_json_snapshot!(error_test("listOfListOfStrings", "[[String!]!]!", json!([["hello", 1, "there"]])), @r###"
+    [
+      "Variable $input got an invalid value: found a Integer value where we expected a String scalar at $input.0.1"
+    ]
+    "###);
+}
+
+#[test]
+fn invalid_input_objects() {
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"string": 1})),
+        @r###"
+    [
+      "Variable $input got an invalid value: found a Integer value where we expected a String scalar at $input.string"
+    ]
+    "###
+    );
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"int": "hello"})),
+        @r###"
+    [
+      "Variable $input got an invalid value: found a String value where we expected a Int scalar at $input.int"
+    ]
+    "###
+    );
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"recursiveObject": {"string": 1}})),
+        @r###"
+    [
+      "Variable $input got an invalid value: found a Integer value where we expected a String scalar at $input.recursiveObject.string"
+    ]
+    "###
+    );
+    // This one is valid because it gets list coerced
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"recursiveObjectList": {}})),
+        @"[]"
+    );
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"recursiveObjectList": [null]})),
+        @r###"
+    [
+      "Variable $input got an invalid value: found a null where we expected a non-null type at $input.recursiveObjectList.0"
+    ]
+    "###
+    );
+    // This one is also valid because it gets list coerced
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"recursiveObjectList": [{"recursiveObjectList": {}}]})),
+        @"[]"
+    );
+    insta::assert_json_snapshot!(
+        error_test("inputObject", "InputObj!", json!({"recursiveObjectList": [{"recursiveObjectList": [null]}]})),
+        @r###"
+    [
+      "Variable $input got an invalid value: found a null where we expected a non-null type at $input.recursiveObjectList.0.recursiveObjectList.0"
+    ]
+    "###
+    );
+}
+
+#[test]
+fn multiple_invalid_variables() {
+    let query = "query($one: String!, $two: Int!) { string(input: $one) int(input: $two) }";
+
+    let errors = run_query(query, &json!({"one": true, "two": "hello"}))
+        .errors()
+        .iter()
+        .map(|error| error["message"].as_str().expect("message to be a string").to_string())
+        .collect::<Vec<_>>();
+
+    insta::assert_json_snapshot!(errors, @r###"
+    [
+      "Variable $one got an invalid value: found a Boolean value where we expected a String scalar at $one",
+      "Variable $two got an invalid value: found a String value where we expected a Int scalar at $two"
+    ]
+    "###);
```

Comment:
Just in case it's an omission, I'm not sure I've seen cases about enum value coercions

Reply:
Ah yes, good point.  I'll add some of those